syntax = "proto2";

package bacs.problem.single.testing;

import "bacs/problem/single/settings.proto";

message SolutionTesting {
  repeated TestGroup test_group = 1;
}

message Dependency {
  enum Requirement {
    ALL_OK = 0;
    ALL_FAIL = 1;
    AT_LEAST_ONE_OK = 2;
    AT_MOST_ONE_FAIL = 3;
    AT_LEAST_HALF_OK = 4;
  }
  required string test_group = 1;
  optional Requirement requirement = 2 [default = ALL_OK];
}

message TestGroup {
  required string id = 1;
  required settings.TestGroupSettings settings = 2;
  repeated TestQuery test_set = 3;
  repeated Dependency dependency = 4;
}

message TestQuery {
  oneof query {
    string id = 1;
    WildcardQuery wildcard = 2;
    RegexQuery regex = 3;
  }
}

message WildcardQuery {
  enum Flag {
    IGNORE_CASE = 0;
  }
  repeated Flag flag = 1;
  required string value = 2;
}

message RegexQuery {
  enum Flag {
    IGNORE_CASE = 0;
  }
  repeated Flag flag = 1;
  required string value = 2;
}
